components/builder-notify/src/server.rs

```diff
@@ -0,0 +1,56 @@
+use crate::{bldr_events::connection::{create_consumer,
+                                      EventBusConsumer},
+            config::Config,
+            error::Error,
+            hab_core::{config::ConfigFile,
+                       ok_warn}};
+use std::{path::PathBuf,
+          thread::{self},
+          time::Duration};
+
+const CFG_DEFAULT_PATH: &str = "/hab/svc/builder-notify/config/config.toml";
+const DEFAULT_QUEUE_NAME: &str = "builder_events";
+
+pub struct AppState {
+    eventconsumer: Option<Box<dyn EventBusConsumer>>,
+}
+
+impl AppState {
+    pub fn new(config: &Config) -> Result<AppState, Error> {
+        let mut app_state = AppState { eventconsumer: None, };
+
+        app_state.eventconsumer = ok_warn!(create_consumer(&config.eventbus));
+
+        Ok(app_state)
+    }
+}
+
+pub fn run(path: Option<PathBuf>) -> Result<(), Error> {
+    info!("Launching builder-notify service.");
+    let config_path = path.unwrap_or_else(|| PathBuf::from(CFG_DEFAULT_PATH));
+    let config = if config_path.is_file() {
+        Config::from_file(config_path).map_err(|e| Error::NotificationsError(Box::new(e)))?
+    } else {
+        warn!("EventBusConsumer config file not found at {:?}. Using defaults..",
+              config_path);
+        Config::default()
+    };
+    match AppState::new(&config) {
+        Ok(state) => {
+            if let Some(bus) = state.eventconsumer {
+                info!("EventBusConsumer started.");
+                bus.subscribe([DEFAULT_QUEUE_NAME].to_vec())
+                   .map_err(|e| Error::NotificationsError(Box::new(e)))?;
+
+                loop {
+                    if let Some(msg) = bus.poll() {
+                        debug!("received msg {:?}", msg);
+                        thread::sleep(Duration::from_millis(100));
```

Comment:
This got me thinking a bit... while this will work, you're still going to be using more CPU than you need. It looks like your `poll` method definition is strongly determined by your choice of the synchronous Kafka consumer. This in turn means that anything else that's going to implement `EventBusConsumer` can't be async, which seems like it could be limiting going forward.

Have you tried using the Kafka [stream consumer](https://github.com/fede1024/rust-rdkafka/blob/master/src/consumer/stream_consumer.rs) in an attempt to make your abstraction fully async? 

(I realize that that is a not-insignificant change, and could very well be a future refinement, given that you're basically just sketching the outlines of this notifications service right now, which is totally fine!)

Reply:
This is a very astute observation. I originally attempted to implement the Stream Consumer but then became somewhat bogged down in the mechanics of tackling a Trait method with a signature that returns a generic stream (for current and future provider types). For that reason and since, as you mentioned, I'm still in the early sketching phases, I chose to defer that implementation to sometime in the near future.

To that end, I've added a `TODO` comment in https://github.com/habitat-sh/builder/pull/1549/commits/261fcc5a3138e104852fa8476e681b947820d042 as a placeholder to circle back and add in an async message stream consumer.